lib/pbench/test/unit/server/test_api_key.py

```diff
@@ -90,4 +92,162 @@ def test_successful_api_key_generation(self, query_get_as, pbench_drb_token):
+    def test_delete_api_key(self, query_as, pbench_drb_token, pbench_drb_api_key):
+
+        # we can find it
+        key = APIKey.query(api_key=pbench_drb_api_key.api_key)
+        assert key.api_key == pbench_drb_api_key.api_key
+        assert key.id == pbench_drb_api_key.id
+
+        query_as(pbench_drb_token, pbench_drb_api_key.id, HTTPStatus.OK)
+        audit = Audit.query()
+        assert len(audit) == 2
+        assert audit[0].id == 1
+        assert audit[0].root_id is None
+        assert audit[0].operation == OperationCode.DELETE
+        assert audit[0].status == AuditStatus.BEGIN
+        assert audit[0].name == "apikey"
+        assert audit[0].object_type == AuditType.API_KEY
+        assert audit[0].object_id is None
+        assert audit[0].object_name is None
+        assert audit[0].user_id == "3"
+        assert audit[0].user_name == "drb"
+        assert audit[0].reason is None
+        assert audit[0].attributes is None
+        assert audit[1].id == 2
+        assert audit[1].root_id == 1
+        assert audit[1].operation == OperationCode.DELETE
+        assert audit[1].status == AuditStatus.SUCCESS
+        assert audit[1].name == "apikey"
+        assert audit[1].object_type == AuditType.API_KEY
+        assert audit[1].object_id is None
+        assert audit[1].object_name is None
+        assert audit[1].user_id == "3"
+        assert audit[1].user_name == "drb"
+        assert audit[1].reason is None
+        assert audit[1].attributes["key_id"] == pbench_drb_api_key.id
+        assert audit[1].attributes["name"] == "drb_key"
+
+        assert APIKey.query(api_key=pbench_drb_api_key.api_key) is None
+
+    def test_unauthorized_delete(
+        self, query_as, pbench_drb_token_invalid, pbench_drb_api_key
+    ):
+
+        response = query_as(
+            pbench_drb_token_invalid,
+            pbench_drb_api_key.id,
+            HTTPStatus.UNAUTHORIZED,
+        )
+        assert response.json == {
+            "message": "User provided access_token is invalid or expired"
+        }
+
+    def test_delete_api_key_notfound(self, query_as, pbench_drb_token):
+        response = query_as(
+            pbench_drb_token, "pbench_drb_token_invalid", HTTPStatus.NOT_FOUND
+        )
+        assert response.json == {"message": "Requested key not found"}
+        audit = Audit.query()
+        assert len(audit) == 2
+        assert audit[0].id == 1
+        assert audit[0].root_id is None
+        assert audit[0].operation == OperationCode.DELETE
+        assert audit[0].status == AuditStatus.BEGIN
+        assert audit[0].name == "apikey"
+        assert audit[0].object_type == AuditType.API_KEY
+        assert audit[1].id == 2
+        assert audit[1].root_id == 1
+        assert audit[1].operation == OperationCode.DELETE
+        assert audit[1].status == AuditStatus.FAILURE
+        assert audit[1].name == "apikey"
+        assert audit[1].object_type == AuditType.API_KEY
+        assert audit[1].attributes is None
```

Comment:
Yes, I was talking about the code which creates the audit record:  currently it records the fact that the creation attempt fails, but it doesn't record why or most of the details of the request.  Admittedly, this is more like a "logging thing" than an "auditing thing", but, since we're going to create the audit entry anyway, it might be worth adding this details there.

I think all that is required is to move the assignment of `context["auditing"]["attributes"]` up near the top of `_post()` so that it is already set if/when a failure occurs.

Reply:
We should treat `AuditType.API_KEY` like a real resource now that we have the `id` column, and `DELETE` should report `object_type = AuditType.API_KEY, object_id = APIKey.id, object_name = APIKey.label`.

We still can't include the ID on a `POST /api/v1/key` because we don't know the ID yet, so I don't think we have much practical alternative but to include that in `attributes`. However for `DELETE` the `BEGIN` audit record should be fully populated. (This would also follow if we did a `PUT /api/vi/key/{id}` to *update* a key, but that would only allow changing the label and at this point I don't think that's worthwhile.)